reactive/flannel.py

```diff
@@ -1,254 +1,218 @@
-    # out of the HA connection string (if applicable)
-    hosts = connection_string.split(',')
-    host = hosts[0].split(':')[1].lstrip('//')
-    port = hosts[0].split(':')[2]
-
-    tls_required = is_state('etcd.tls.available')
-    if tls_required:
-        cert_path = '/etc/ssl/flannel'
-        key = '{}/client-key.pem'.format(cert_path)
-        cert = '{}/client-cert.pem'.format(cert_path)
-        ca = '{}/client-ca.pem'.format(cert_path)
-        client = etcd.Client(host=host, port=int(port), protocol='https',
-                             cert=(cert, key), ca_cert=ca)
+            default_interface = line.split(' ')[-1]
+            break
+    context = {'iface': config('iface') or default_interface,
+               'connection_string': etcd.get_connection_string(),
+               'cert_path': ETCD_PATH}
+    render('flannel.service', '/lib/systemd/system/flannel.service', context)
+    set_state('flannel.service.installed')
+    remove_state('flannel.service.started')
+
+
+@when('config.changed.iface')
+def reconfigure_flannel_service():
+    ''' Handle interface configuration change. '''
+    remove_state('flannel.service.installed')
+
+
+@when('flannel.binaries.installed', 'flannel.etcd.credentials.installed',
+      'etcd.available')
+@when_not('flannel.network.configured')
+def configure_network(etcd):
+    ''' Store initial flannel data in etcd. '''
+    data = json.dumps({
+        'Network': config('cidr'),
+        'Backend': {
+            'Type': 'vxlan'
+        }
+    })
+    cmd = "etcdctl "
+    cmd += "--endpoint '{0}' ".format(etcd.get_connection_string())
+    cmd += "--cert-file {0} ".format(ETCD_CERT_PATH)
+    cmd += "--key-file {0} ".format(ETCD_KEY_PATH)
+    cmd += "--ca-file {0} ".format(ETCD_CA_PATH)
+    cmd += "set /coreos.com/network/config '{0}'".format(data)
+    check_call(split(cmd))
+    set_state('flannel.network.configured')
+    remove_state('flannel.service.started')
+
+
+@when('config.changed.cidr')
+def reconfigure_network():
+    ''' Trigger the network configuration method. '''
+    remove_state('flannel.network.configured')
+
+
+@when('flannel.binaries.installed', 'flannel.service.installed',
+      'flannel.network.configured')
+@when_not('flannel.service.started')
+def start_flannel_service():
+    ''' Start the flannel service. '''
+    status_set('maintenance', 'Starting flannel service.')
+    if service_running('flannel'):
+        service_restart('flannel')
     else:
-        client = etcd.Client(host=host, port=int(port), protocol='http')
+        service_start('flannel')
+    set_state('flannel.service.started')
 
-    data = {'Network': hookenv.config('cidr'), 'Backend': {'Type': 'vxlan'}}
-    json_data = json.dumps(data)
-    client.write('/coreos.com/network/config', json_data)
 
+@when('flannel.service.started', 'cni.is-worker')
+@when_not('flannel.cni.available')
+def set_available(cni):
+    ''' Indicate to the CNI provider that we're ready. '''
+    cni.set_available()
```

Comment:
This just signals that CNI is ready to go. Kubernetes-worker waits for this signal before starting kubelet.

CNI is ready to go when:
1. CNI configuration has been created in `/etc/cni/net.d/`
2. All necessary CNI plugins have been installed to `/opt/cni/bin/`

Having said that, it looks like the decorators for this step are wrong. Looking into it.